### PR TITLE
Optimise catalog function rebase of #3145

### LINKED ISSUE #3233: OCI Conformance: GET start of tag is set by `last` query parameter
epic: https://github.com/docker/distribution/issues/3203

### Content Discovery > Listing tags

**GET number of tags should be limitable by `n` query parameter**

To resolve this issue, please see the open pull request [3145](https://github.com/docker/distribution/pull/3145). This must be resolved after the preceding issue in order to return consistent results. Once results are sorted lexically, the `last` query parameter will specify the start (noninclusive) of the subset of `n` tags to be returned.

```
2020/07/10 06:16:25.230803 DEBUG
==============================================================================
~~~ REQUEST ~~~
GET  /v2/oci-conformance/distribution-test/tags/list?n=2  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 200 OK
RECEIVED AT  : 2020-07-10T06:16:25.230528783Z
TIME DURATION: 1.260204ms
HEADERS      :
	Content-Length: 86
	Content-Type: application/json
	Date: Fri, 10 Jul 2020 06:16:25 GMT
	Docker-Distribution-Api-Version: registry/2.0
	X-Content-Type-Options: nosniff
BODY         :
{
   "name": "oci-conformance/distribution-test",
   "tags": [
      "test1",
      "test0",
      "test2",
      "test3"
   ]
}

==============================================================================

2020/07/10 06:16:25.232259 DEBUG
==============================================================================
~~~ REQUEST ~~~
GET  /v2/oci-conformance/distribution-test/tags/list?last=test0&n=2  HTTP/1.1
HOST   : distribution:5000
HEADERS:
	User-Agent: distribution-spec-conformance-tests
BODY   :
***** NO CONTENT *****
------------------------------------------------------------------------------
~~~ RESPONSE ~~~
STATUS       : 200 OK
RECEIVED AT  : 2020-07-10T06:16:25.231990686Z
TIME DURATION: 1.130003ms
HEADERS      :
	Content-Length: 86
	Content-Type: application/json
	Date: Fri, 10 Jul 2020 06:16:25 GMT
	Docker-Distribution-Api-Version: registry/2.0
	X-Content-Type-Options: nosniff
BODY         :
{
   "name": "oci-conformance/distribution-test",
   "tags": [
      "test1",
      "test0",
      "test2",
      "test3"
   ]
}

==============================================================================

Expected
    <int>: 4
to be <=
    <int>: 2
```

### PR DESCRIPTION
**Pull request by DavidSpek**
Rebase of https://github.com/distribution/distribution/pull/3145 to sign off commits and unblock v3 release.

Fixes: https://github.com/distribution/distribution/pull/3145

> fixes #3233 OCI Conformance: GET start of tag is set by last query parameter
> 
> Basically this implementation will only stat `n` times, and only look at the dirs required. E.g. with around 6.500 images, this will only call the `rec` function ~100 times (which also means stat that amount), whereas the other will be called ~20.000 times, where it will stat them all.
> 
> This also fixes a bug where if you specify something close to a last entry, e.g. `fo`, and `foo` is meant, then it will work. This should not be the behavior, as it can cause various bugs.
> 
> Removed the `lessPath` as it isn't needed anymore. Fun fact, the non-native method is slower than native.
> 
> ```
> BenchmarkPathCompareEqual-4         	 3906804	         297 ns/op
> BenchmarkPathCompareNotEqual-4      	148567686	         7.93 ns/op
> BenchmarkPathCompareNative-4        	212109276	         5.61 ns/op
> BenchmarkPathCompareNativeEqual-4   	283666616	         3.83 ns/op
> ```
> 
> Should be ready to merge!

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204651419195143

### CODE REVIEW

**Comment by DavidSpek** [comment]
@milosgajdos since you reviewed the original PR maybe you can have a look at this one that has been rebased and the commits are signed off so it can be merged (if the code is good).

**Comment by DavidSpek** [comment]
Friendly ping @thaJeztah @squizzi

**Comment by milosgajdos** [comment]
I was one of the co-authors of the original PR, so I'd like another pair of eyes on this; I dont wanna review my own code, we all know how that can go 🫠

**Comment by DavidSpek** [comment]
Yeah that isn't great. Anybody else that might be able to review this?

**Comment by milosgajdos** [comment]
PTAL @squizzi @wy65701436

**Comment by DavidSpek** [comment]
Friendly ping @thaJeztah @squizzi @wy65701436. This PR should finalize OCI conformance and close 3 milestones for the v3 release.

**Comment by DavidSpek** [comment]
@deleteriousEffect Would you maybe be so kind as to give this PR a review?

**Comment by DavidSpek** [comment]
@chrispat Maybe you can also provide a review for this PR.

File: registry/storage/catalog.go
Comment (deleteriousEffect) [inline registry/storage/catalog.go:29]:
```suggestion
	if len(repos) == 0 && cap(repos) == 0 {
```

Otherwise, we'll error out on zero length slices that have a capacity such as one initialized like `var s = make([]string, 0, 100)`

File: registry/storage/catalog.go
Comment (deleteriousEffect) [inline registry/storage/catalog.go:30]:
```suggestion
		return 0, errors.New("no repos requested")
```

We're already returning an error, we don't need to also return a C-like special case `-1` here as well.

File: registry/storage/catalog_test.go
Comment (deleteriousEffect) [inline registry/storage/catalog_test.go:242]: We need to add a new test to demonstrate that the bug has been fixed.

**Review by deleteriousEffect — CHANGES_REQUESTED** [review]
I have some feedback, nothing structural, but this is a pretty significant change in implementation, so I feel a little more cautious than normal.

This also affects the garbage collector as well, as it uses the repository enumerator to iterate over repositories. That's a destructive action, so we want to have parity aside from the fix that's part of this change.

The upload purger isn't affected, but I'm curious now if that's the only thing left using the `Walk` function. If so, it might be nice to use the same technique there and get rid of the `Walk` implementations completely, since they're so complex for only having a single caller.

File: registry/storage/catalog_test.go
Comment (DavidSpek) [reply registry/storage/catalog_test.go:242]: Would you have any suggestions? I'm not actually familiar with the code base, I just rebased the PR so we can get it merged to move the v3 release forward.

Comment (DavidSpek) [reply registry/storage/catalog_test.go:242]: Or maybe @milosgajdos can way in here since you worked on the original PR.

Comment (milosgajdos) [reply registry/storage/catalog_test.go:242]: my memory is fading 😄 but I'll try to have a look

**Comment by DavidSpek** [comment]
@wy65701436  Can you also have a look at this.

File: registry/storage/catalog.go
Comment (milosgajdos) [inline registry/storage/catalog.go:30]: Is there a reason why we're checking for slice capacity here? 🤔

File: registry/storage/catalog_test.go
Comment (milosgajdos) [reply registry/storage/catalog_test.go:242]: Just FYI @deleteriousEffect, that PR _optimises_ the catalog iteration. It does not fix any bugs per se: https://github.com/distribution/distribution/issues/3233#issuecomment-1357731317

The original PR shows benchmarks; I dont think we should keep blocking this PR 🤔

File: registry/storage/catalog.go
Comment (deleteriousEffect) [reply registry/storage/catalog.go:30]: https://github.com/distribution/distribution/pull/3902#discussion_r1242771306

File: registry/storage/catalog_test.go
Comment (deleteriousEffect) [reply registry/storage/catalog_test.go:242]: > It does not fix any bugs per se

If it doesn't fix any bugs that's alright, but the PR title and commit message needs to reflect that.

**Comment by milosgajdos** [comment]
@DavidSpek can you please change the title of this PR - it's confusing as this is not addressing any bugs

File: registry/storage/catalog_test.go
Comment (milosgajdos) [reply registry/storage/catalog_test.go:242]: > the PR title and commit message needs to reflect that.

Fair. I changed the PR title. Updating commit messages is tricky as this is a rebase of the abandoned PR. We could squash them, perhaps and give credit to the original contribution in the squashed commit. WDYT?

**Comment by DavidSpek** [comment]
Sorry, I was away on vacation. Looks like the PR title has been updated already.

File: registry/storage/catalog_test.go
Comment (DavidSpek) [reply registry/storage/catalog_test.go:242]: @milosgajdos I've just changed the commit message so it doesn't mention a bug fix anymore.

**Comment by DavidSpek** [comment]
@milosgajdos @deleteriousEffect Can we try and get this over the line soon?

**Review by deleteriousEffect — APPROVED** [review]

**Review by joaodrp — APPROVED** [review]
LGTM!

**Comment by eyJhb** [comment]
Thanks for getting this fixed up and merged <3 Highly appreciated !

**Comment by eyJhb** [comment]
Curious, as I can't seem to find it. Was the CodeQL ever fixed? Ie. potential path traversal. I can't seem to find any prevention for this. With was originally why the other PR went stale :)

**Comment by milosgajdos** [comment]
It was indeed. We wouldnt merge if it werent 😄 Mind you there were also bugs fixed in CodeQL itself, too.

Screenshot 2023-08-18 at 08 58 05

**Comment by eyJhb** [comment]
Hmm, I see no code changes related to it. Maybe CodeQL just got smarter then :p

**Pull request merged**
